From 2a9060b4b6670ad4db169abeac17728f56b43139 Mon Sep 17 00:00:00 2001
From: Marco Eichelberg <eichelberg@offis.de>
Date: Wed, 25 Mar 2026 15:20:19 +0100
Subject: [PATCH] Adapted DCMTK for use with OpenSSL 4.0.0

Added a new configure test and updated dcmsign code to work correctly
with OpenSSL 4.0.0.

Upstream: https://github.com/DCMTK/dcmtk/commit/2a9060b4b6670ad4db169abeac17728f56b43139
(cherry picked from commit 2a9060b4b6670ad4db169abeac17728f56b43139)
---
 CMake/dcmtkPrepare.cmake   |  1 +
 CMake/osconfig.h.in        |  4 ++++
 dcmsign/libsrc/sicert.cc   | 43 ++++++++++++++++++++++++++++++++++----
 dcmsign/libsrc/sitstamp.cc | 19 +++++------------
 4 files changed, 49 insertions(+), 18 deletions(-)

diff --git a/CMake/dcmtkPrepare.cmake b/CMake/dcmtkPrepare.cmake
index 2e9e92271..aaf817b52 100644
--- a/CMake/dcmtkPrepare.cmake
+++ b/CMake/dcmtkPrepare.cmake
@@ -712,6 +712,7 @@ if(DCMTK_WITH_OPENSSL)
   CHECK_FUNCTIONWITHHEADER_EXISTS("TLS1_TXT_ECDHE_ECDSA_WITH_AES_256_CCM_8" "openssl/ssl.h" HAVE_OPENSSL_PROTOTYPE_TLS1_TXT_ECDHE_ECDSA_WITH_AES_256_CCM_8)
   CHECK_FUNCTIONWITHHEADER_EXISTS("TLS1_TXT_ECDHE_ECDSA_WITH_CAMELLIA_256_GCM_SHA384" "openssl/ssl.h" HAVE_OPENSSL_PROTOTYPE_TLS1_TXT_ECDHE_ECDSA_WITH_CAMELLIA_256_GCM_SHA384)
   CHECK_FUNCTIONWITHHEADER_EXISTS("TS_VERIFY_CTX_set0_store(0,0)" "openssl/ts.h" HAVE_OPENSSL_PROTOTYPE_TS_VERIFY_CTX_SET0_STORE)
+  CHECK_FUNCTIONWITHHEADER_EXISTS("X509_check_certificate_times(0,0,0)" "openssl/x509.h" HAVE_OPENSSL_PROTOTYPE_X509_CHECK_CERTIFICATE_TIMES)
 
   # test presence of functions, constants and macros needed for the dcmsign module
   CHECK_FUNCTIONWITHHEADER_EXISTS("EVP_PKEY_get_group_name" "openssl/evp.h" HAVE_OPENSSL_PROTOTYPE_EVP_PKEY_GET_GROUP_NAME)
diff --git a/CMake/osconfig.h.in b/CMake/osconfig.h.in
index 6d39b4e40..44aab5ab5 100644
--- a/CMake/osconfig.h.in
+++ b/CMake/osconfig.h.in
@@ -773,9 +773,13 @@ DCMTK was configured to use C++20 features, but your compiler does not or was no
 /* Define if we have OpenSSL with the OSSL_PROVIDER_load() function */
 #cmakedefine HAVE_OPENSSL_PROTOTYPE_OSSL_PROVIDER_LOAD @HAVE_OPENSSL_PROTOTYPE_OSSL_PROVIDER_LOAD@
 
+/* Define if we have OpenSSL with the X509_check_certificate_times() function */
+#cmakedefine HAVE_OPENSSL_PROTOTYPE_X509_CHECK_CERTIFICATE_TIMES @HAVE_OPENSSL_PROTOTYPE_X509_CHECK_CERTIFICATE_TIMES@
+
 /* Define if we have the <openssl/provider.h> header file*/
 #cmakedefine HAVE_OPENSSL_PROVIDER_H @HAVE_OPENSSL_PROVIDER_H@
 
+
 /* Historical C/C++ language features and APIs that are guaranteed by C++98,
  * C99, or POSIX.1-2001 for Posix platforms, or are not used anywhere in DCMTK,
  * and thus not tested anymore.
diff --git a/dcmsign/libsrc/sicert.cc b/dcmsign/libsrc/sicert.cc
index 0668b72c9..7cd5f4a07 100644
--- a/dcmsign/libsrc/sicert.cc
+++ b/dcmsign/libsrc/sicert.cc
@@ -1,6 +1,6 @@
 /*
  *
- *  Copyright (C) 1998-2024, OFFIS e.V.
+ *  Copyright (C) 1998-2026, OFFIS e.V.
  *  All rights reserved.  See COPYRIGHT file for details.
  *
  *  This software and supporting documentation were developed by
@@ -38,6 +38,8 @@ BEGIN_EXTERN_C
 #include <openssl/x509.h>
 #include <openssl/pem.h>
 #include <openssl/err.h>
+#include <openssl/asn1.h>
+#include <openssl/x509_vfy.h>
 END_EXTERN_C
 
 
@@ -352,6 +354,22 @@ OFBool SiCertificate::isCertExpiredNow() const
   OFBool result = OFTrue;
   if (x509)
   {
+#ifdef HAVE_OPENSSL_PROTOTYPE_X509_CHECK_CERTIFICATE_TIMES
+    /* X509_check_certificate_times() was introduced in OpenSSL 4.0.0 */
+    int error = 0;
+    X509_VERIFY_PARAM *param = X509_VERIFY_PARAM_new();
+    if (param)
+    {
+        // return false if the certificate is valid, or if it invalid for a reason
+        // not related to the expiry date
+        if (X509_check_certificate_times(param, x509, &error)
+           || ((error != X509_V_ERR_CERT_HAS_EXPIRED)
+           && (error != X509_V_ERR_ERROR_IN_CERT_NOT_AFTER_FIELD)))
+           result = OFFalse;
+        X509_VERIFY_PARAM_free(param);
+    }
+#else
+
     const ASN1_TIME *certexpiry = X509_get0_notAfter(x509);
     if (certexpiry)
     {
@@ -359,6 +377,7 @@ OFBool SiCertificate::isCertExpiredNow() const
       // and zero if there is an error. In both cases we treat the certificate as expired.
       if (X509_cmp_current_time(certexpiry) > 0) result = OFFalse;
     }
+#endif
   }
   return result;
 }
@@ -369,6 +388,21 @@ OFBool SiCertificate::isCertNotYetValidNow() const
   OFBool result = OFTrue;
   if (x509)
   {
+#ifdef HAVE_OPENSSL_PROTOTYPE_X509_CHECK_CERTIFICATE_TIMES
+    /* X509_check_certificate_times() was introduced in OpenSSL 4.0.0 */
+    int error = 0;
+    X509_VERIFY_PARAM *param = X509_VERIFY_PARAM_new();
+    if (param)
+    {
+        // return false if the certificate is valid, or if it invalid for a reason
+        // not related to the expiry date
+        if (X509_check_certificate_times(param, x509, &error)
+           || ((error != X509_V_ERR_CERT_NOT_YET_VALID)
+           && (error != X509_V_ERR_ERROR_IN_CERT_NOT_BEFORE_FIELD)))
+           result = OFFalse;
+        X509_VERIFY_PARAM_free(param);
+    }
+#else
     const ASN1_TIME *certstart = X509_get0_notBefore(x509);
     if (certstart)
     {
@@ -376,6 +410,7 @@ OFBool SiCertificate::isCertNotYetValidNow() const
       // and zero if there is an error. In both cases we treat the certificate as invalid.
       if (X509_cmp_current_time(certstart) < 0) result = OFFalse;
     }
+#endif
   }
   return result;
 }
@@ -511,9 +546,9 @@ OFCondition SiCertificate::convertGeneralizedTime(const ASN1_GENERALIZEDTIME *d,
     int n, i, l, o;
 
     if (d == NULL) return EC_IllegalCall;
-    if (d->type != V_ASN1_GENERALIZEDTIME) return EC_IllegalCall;
-    l = d->length;
-    a = (char *)d->data;
+    if (ASN1_STRING_type(d) != V_ASN1_GENERALIZEDTIME) return EC_IllegalCall;
+    l = ASN1_STRING_length(d);
+    a = (char *)ASN1_STRING_get0_data(d);
     o = 0;
     /*
      * GENERALIZEDTIME is similar to UTCTIME except the year is represented
diff --git a/dcmsign/libsrc/sitstamp.cc b/dcmsign/libsrc/sitstamp.cc
index 7305d3be5..bab018f00 100644
--- a/dcmsign/libsrc/sitstamp.cc
+++ b/dcmsign/libsrc/sitstamp.cc
@@ -1,6 +1,6 @@
 /*
  *
- *  Copyright (C) 2019-2024, OFFIS e.V.
+ *  Copyright (C) 2019-2026, OFFIS e.V.
  *  All rights reserved.  See COPYRIGHT file for details.
  *
  *  This software and supporting documentation were developed by
@@ -274,24 +274,14 @@ OFCondition SiTimeStamp::create_ts_query(
 
         if (nonce_asn1)
         {
-          // free memory buffer of nonce_asn1 (we will create a new one)
-          OPENSSL_free(nonce_asn1->data);
 
           // Find the first non-zero byte in the buffer
           int i = 0;
           for (i = 0; i < NONCE_LENGTH && !nonce_buf[i]; ++i)
             continue;
 
-          // create a new memory buffer for nonce_asn1
-          nonce_asn1->length = NONCE_LENGTH - i;
-          nonce_asn1->data = OFreinterpret_cast(unsigned char *, OPENSSL_malloc(nonce_asn1->length + 1));
-          if (NULL != nonce_asn1->data)
-          {
-              // copy random data into nonce. The first byte is guaranteed to be nonzero,
-              // which is necessary because otherwise the field might violate DER encoding.
-              memcpy(nonce_asn1->data, nonce_buf + i, nonce_asn1->length);
-          }
-          else result = EC_MemoryExhausted;
+          if (0 == ASN1_STRING_set(nonce_asn1, nonce_buf + i, NONCE_LENGTH - i))
+              result = EC_MemoryExhausted;
 
           if (result.good())
           {
@@ -1071,7 +1061,8 @@ void SiTimeStamp::get_tsinfo_extension(OFString& ext, int idx) const
     }
     else
     {
-      X509_EXTENSION *x509ext = TS_TST_INFO_get_ext(tsinfo_, idx);
+      // need to cast the const qualifier away for 1.x OpenSSL versions
+      X509_EXTENSION *x509ext = OFconst_cast(X509_EXTENSION *, TS_TST_INFO_get_ext(tsinfo_, idx));
       if (x509ext == NULL)
       {
         DCMSIGN_WARN("timestamp extension " << idx << " cannot be accessed");
